src/garage/sampler/pearl_sampler.py

```diff
@@ -0,0 +1,73 @@
+# pylint: disable=unnecessary-pass
+"""A sampler that does not perform serialization for sampling."""
+import numpy as np
+
+from garage.sampler.utils import rollout
+
+
+class PEARLSampler:
+    """A sampler used in meta-RL algorithms involving context.
+
+    It stores context and resample belief in the policy every step.
+
+    Args:
+        env (garage.envs.GarageEnv): An environement instance.
+        policy (garage.policies.Policy): Policy used for sampling.
+        max_path_length (int): Maximum length of path for each step
+
+    """
+
+    def __init__(self, env, policy, max_path_length):
+        self.env = env
+        self.policy = policy
+        self.max_path_length = max_path_length
+
+    def start_worker(self):
+        """Start worker."""
+        pass
+
+    def shutdown_worker(self):
+        """Terminate worker."""
+        pass
+
+    def obtain_samples(self,
+                       max_samples=np.inf,
+                       max_trajs=np.inf,
+                       accum_context=True,
+                       resample=1):
+        """Obtain samples in the environment up to max_samples or max_trajs.
+
+        Args:
+            max_samples (int): Maximum number of samples.
+            max_trajs (int):  Maximum number of trajectories.
+            accum_context (bool): Whether or not to accumulate the collected
+                context.
+            resample (int): How often (in trajectories) to resample context.
+
+        Returns:
+            list: A list of paths.
+            int: Total number of steps.
+
+        """
+        assert max_samples < np.inf or max_trajs < np.inf
+        paths = []
+        n_steps_total = 0
+        n_trajs = 0
+
+        while n_steps_total < max_samples and n_trajs < max_trajs:
+            path = rollout(self.env,
+                           self.policy,
+                           max_path_length=self.max_path_length,
```

Comment:
Do we ever want max path length to be infinite? This would make it such that there's never a reset in some environments that definitely need a reset after a certain number of steps. For example, standard mujoco environments should have max path length set to 1000.

Reply:
i'm not sure 

> 1. Based on comments and code similarities, not sure if we need a new meta-rl replay buffer. Perhaps a refactor using an existing replay buffer can come later if we decide.
> 2. What is the reasoning behind a multi task replay buffer for PEARL? I saw your comment:
> 
> > The algorithm keeps a buffer for each task, and I think this looks cleaner. I just followed the implementation on oyster, but I suppose I can keep a list of buffers in the algo without this buffer class.
> 
> I'm with you on keeping a list of buffers as opposed to a separate MT replay buffer
> 
> 1. Can you create your own rollout function separate from the sampler/utils `rollout` function.
>    It seems that you have modified the return types of the `rollout function` and added some other features. A question I find myself asking is "Will the added functionality be used by algorithms besides PEARL?"

Let's keep the major changes to a minimum here. We can make better changes once everything is merged.

Since many of the ways of improving this code involve unifying it with other algorithms, it is better to merge high-quality working code than to insist that it be high-quality, working, and unified.

Reply:
Kk

Reply:
Kk